Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -432,94 +431,94 @@ void GCMemcardManager::FixChecksums()
   }
 
   return QPixmap::fromImage(image);
 }
 
-std::vector<QPixmap> GCMemcardManager::GetIconFromSaveFile(int file_index, int slot)
+GCMemcardManagerIconAnimationData GCMemcardManager::GetIconFromSaveFile(int file_index, int slot)
 {
   auto& memcard = m_slot_memcard[slot];
 
-  std::vector<u32> pxdata(BANNER_WIDTH * IMAGE_HEIGHT);
   std::vector<u8> anim_delay(ANIM_MAX_FRAMES);
-  std::vector<u32> anim_data(ANIM_FRAME_WIDTH * IMAGE_HEIGHT * ANIM_MAX_FRAMES);
+  std::vector<u32> anim_data(ICON_WIDTH * ICON_HEIGHT * ANIM_MAX_FRAMES);
 
-  std::vector<QPixmap> frame_pixmaps;
+  GCMemcardManagerIconAnimationData frame_data;
 
   u32 num_frames = memcard->ReadAnimRGBA8(file_index, anim_data.data(), anim_delay.data());
+  bool is_pingpong = memcard->DEntry_IsPingPong(file_index);
 
   // Decode Save File Animation
   if (num_frames > 0)
   {
-    u32 frames = BANNER_WIDTH / ANIM_FRAME_WIDTH;
-
-    if (num_frames < frames)
+    frame_data.m_frames.reserve(num_frames);
+    const u32 per_frame_offset = ICON_WIDTH * ICON_HEIGHT;
+    for (u32 f = 0; f < num_frames; ++f)
     {
-      frames = num_frames;
-
-      // Clear unused frame's pixels from the buffer.
-      std::fill(pxdata.begin(), pxdata.end(), 0);
+      QImage img(reinterpret_cast<u8*>(&anim_data[f * per_frame_offset]), ICON_WIDTH, ICON_HEIGHT,
+                 QImage::Format_ARGB32);
+      frame_data.m_frames.push_back(QPixmap::fromImage(img));
+      for (int i = 0; i < anim_delay[f]; ++i)
+      {
+        frame_data.m_frame_timing.push_back(static_cast<u8>(f));
+      }
     }
 
-    for (u32 f = 0; f < frames; ++f)
+    if (is_pingpong && num_frames >= 3)
     {
-      for (u32 y = 0; y < IMAGE_HEIGHT; ++y)
+      // if the animation 'ping-pongs' between start and end then the animation frame order is
+      // something like 'abcdcbabcdcba' instead of the usual 'abcdabcdabcd'
+      // to display that correctly just append all except the first and last frame in reverse order
+      // at the end of the animation
+      for (u32 f = num_frames - 2; f > 0; --f)
       {
-        for (u32 x = 0; x < ANIM_FRAME_WIDTH; ++x)
+        for (int i = 0; i < anim_delay[f]; ++i)
         {
-          // NOTE: pxdata is stacked horizontal, anim_data is stacked vertical
-          pxdata[y * BANNER_WIDTH + f * ANIM_FRAME_WIDTH + x] =
-              anim_data[f * ANIM_FRAME_WIDTH * IMAGE_HEIGHT + y * IMAGE_HEIGHT + x];
+          frame_data.m_frame_timing.push_back(static_cast<u8>(f));
         }
       }
     }
-    QImage anims(reinterpret_cast<u8*>(pxdata.data()), BANNER_WIDTH, IMAGE_HEIGHT,
-                 QImage::Format_ARGB32);
-
-    for (u32 f = 0; f < frames; f++)
-    {
-      frame_pixmaps.push_back(
-          QPixmap::fromImage(anims.copy(ANIM_FRAME_WIDTH * f, 0, ANIM_FRAME_WIDTH, IMAGE_HEIGHT)));
-    }
   }
   else
   {
     // No Animation found, use an empty placeholder instead.
-    frame_pixmaps.push_back(QPixmap());
+    frame_data.m_frames.reserve(1);
```

Comment:
The two reserves here aren't necessary if the vectors are empty and you're appending a single entry, given it wouldn't have even allocated memory anyways at that point, so the behavior would be equivalent.